modules/peridynamics/include/kernels/MechanicsBaseNOSPD.h

```diff
@@ -0,0 +1,48 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#ifndef MECHANICSBASENOSPD_H
+#define MECHANICSBASENOSPD_H
+
+#include "MechanicsBasePD.h"
+
+class MechanicsBaseNOSPD;
+class RankTwoTensor;
+class RankFourTensor;
+
+template <>
+InputParameters validParams<MechanicsBaseNOSPD>();
+
+/**
+ * Base kernel class for bond-associated correspondence material models
+ */
+class MechanicsBaseNOSPD : public MechanicsBasePD
+{
+public:
+  MechanicsBaseNOSPD(const InputParameters & parameters);
+
+protected:
+  /**
+   * Function to compute derivative of stress with respect to displacements
+   */
+  virtual RankTwoTensor computeDSDU(unsigned int component, unsigned int nd);
+
+  ///@{ Material point based material properties
+  const MaterialProperty<Real> & _multi;
```

Comment:
@hchen139 Not able to figure out what 'multi' does (?)

Reply:
It is a "factor" for the force state and its value is calculated in material class. Its value is calculated as: weight function * v_i * v_j * bond_associated_horizon_volume / horizon_volume.

Reply:
Ok got it, so it is a factor for the irregular discretization I'm guessing ?

Reply:
This factor is not specific to irregular discretization. But the expression, weight function * v_i * v_j * bond_associated_horizon_volume / horizon_volume, is specific to bond-associated correspondence formulation. For conventional correspondence formulation, it is "weight function * v_i * v_j".